### PR TITLE
Make Ident.t abstract and immutable.

### PR DESCRIPTION
**Pull request by Drup**
The initial version of this patch made Ident.t abstract to avoid exposing the mutability, but I realized that the mutability was not really used anywhere anymore (it used to be important for the toplevel), so I made it immutable as well.

This is useful for compiler forks that play tricks with Idents, in particular the one that have additional flags like *cough* Eliom *cough* and MetaOCaml.
With this, one can modify some functions inside Ident in a way that will not break with future code in other parts of the compiler.

This change is technically breaking. For example, js_of_ocaml's toplevel used to access idents's fields directly (this is not the case anymore). I'm not sure how much it is used otherwise (I guess not much).

### CODE REVIEW

**Review by mshinwell — APPROVED** [review]

**Comment by mshinwell** [comment]
I made a couple of minor edits.  This is a good change.

File: middle_end/base_types/mutable_variable.ml [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 include Variable
 
+<<<<<<< HEAD
```
Comment (alainfrisch) [inline middle_end/base_types/mutable_variable.ml]: Left-over of a conflict marker.

Comment (alainfrisch) [reply middle_end/base_types/mutable_variable.ml]: @mshinwell : careful, I think you merged a broken commit!

**Comment by alainfrisch** [comment]
I've pushed a tentative fix for the build (7110e3e4f).

**Comment by shindere** [comment]
Alain Frisch (2018/04/09 02:43 -0700):
> I've pushed a tentative fix for the build (7110e3e4f).

Thanks. Did that go through AppVeyor before being merged?
If not, why?

**Comment by Drup** [comment]
Thanks for the fixes @mshinwell , but please let me rebase before merging next time. :)

**Comment by mshinwell** [comment]
Sorry, I thought I had built and tested this, but was mistaken.

**Comment by Drup** [comment]
@mshinwell Even then, merge commits in feature branches make the commit history very hard to read and are often unnecessary. For such small patchset, rebasing is both easier and produces a clean history.
Please let me rebase my patchsets before merging them.

**Pull request merged**